woocommerce/woocommerce-android
# Stats Improvements Step 1: Added UI to support the new stats v4 api

Fixes #1259. This PR completes the first step of the Stats Improvements Project as defined in the master thread [here](https://github.com/woocommerce/woocommerce-android/issues/1199). 

### Changes:
- [x] Adds a new FluxC hash that includes the latest v4 api changes.
- [x] Checks if v4 stats api is supported:
  - [x] Adds a new `RevenueStatsAvailabilityFetcher` class that handles the logic to fetch the stats v4 api availability for a site and stores the availability flag in `SharedPreferences`.
  - [x] The availability check is added to the `MainActivity` class and is called every time the app is opened.
  - [x] The availability check is added to the `SitePickerActivity` class and is called every time the user switches a store or picks a store after logging to the app.
  - [x] The availability check is added to the pull to refresh gesture in the old and the new stats. 
- [x] Duplicates the current Dashboard architecture for displaying the new Stats data - This include s duplicating the existing implementation of `DashboardFragment`, `DashboardPresenter`, `DashboardContract`, `DashboardStatsView`, `DashboardModule`. The new classes are added to a separate package called `mystore`.
- [x] The TAB bar title for the v4 stats UI has been changed to `Today`, `This Week`, `This Month`, and `This Year`.
- [x] Since the date ranges have been modified, the date range is passed to FluxC when querying for the visitor stats in v4 UI. 
- [x] Remove the top selling products tab layout.
- [x] Anchor the tabLayout when the view is scrolled.

 ### Notes:
- There is no logic in place, at the moment, to change the UI, if it is already visible to the user. This will be handled in a separate PR. For instance:
  - If the api response returned is successful but the user has already seen the v3 stats screen.
  - If the api response returned is error but the user has already seen the v4 stats screen (happens when the user has disabled this plugin). 
- The v4 stats api does not provide the currency for the data in the api response. So, the logic is modified to fetch the currency from the `Site Settings`.

### Behaviour:
<img width="300" src="https://user-images.githubusercontent.com/22608780/62352407-b2d17680-b525-11e9-8f70-ff0a3295d2b4.gif">

### Screenshots:
#### No Data UI
(LEFT: old stats data . RIGHT: v4 stats data):
<img width="300" src="https://user-images.githubusercontent.com/22608780/62279898-96233900-b468-11e9-8bfa-abfe3e7ec86e.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62352454-d4caf900-b525-11e9-85a3-ffbf90a2ca38.png"> 

#### V4 Stats UI
<img width="300" src="https://user-images.githubusercontent.com/22608780/62352527-fc21c600-b525-11e9-8b6b-aaa053682be6.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62352531-ff1cb680-b525-11e9-8716-a920aa9aa04b.png"> 
<img width="300" src="https://user-images.githubusercontent.com/22608780/62352574-1a87c180-b526-11e9-8525-a94452aa1c3c.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62352578-1cea1b80-b526-11e9-8cab-7a5353b79614.png"> 

### Testing
- Run `MyStorePresenterTest`.
- Run `DateUtilsTest`
- Test a new order completion and verify that the Stats is updated for the current day.
- Test on site without `wc-admin` installed/activated.
- Test on site with `wc-admin` installed.
- Test on site without `wc-admin`, then install `wc-admin` and try again for the same site. The app will not refresh automatically. We would need to close the app and open again to verify the changes.

### Questions:
~~I accidentally changed the style of `TabLayout` so now there is an elevation added to the old stats UI as well. I thought it looked good so I left it that way 😊. We could certainly remove it without much hassle. But WDYT?~~ Removing the elevation for the old stats since there are two tabs and it looks confusing.
~~I noticed while testing that since today is the first of August and there are no orders for my test site today, that:~~
  - ~~Today TAB: has 0 stats.~~
  - ~~This week TAB: has some data (since the stats data is from the start of this week i.e. 29 July).~~
  - ~~This Month TAB: has no data (since the stats data is from the start of the current month i.e. today).~~
  - ~~This Year TAB: has data (since the stats data is from the start of this year)~~

~~I'm not sure if the difference between the data in WEEK and MONTH tab, would be confusing to the user. 😕~~ As @nbradbury mentioned [here](https://github.com/woocommerce/woocommerce-android/pull/1314#issuecomment-517279821), this might not be a problem since we would be adding the date bar to the tabs in another PR.
